### ISSUE TITLE
Cloud sync of settings

### ISSUE THREAD

**Issue by karma1428** [OlegWock/anori#47]

**Comment by mileshc1** [comment]
> 

I would love to see this cloud sync feature, as adding a to-do list on one pc and then remembering to export and import onto my laptop is not manageable. Would be awesome to have some sort of cloud sync.

**Comment by kiznick** [comment]
What if we give a list of free database services (some providers give a free tier and unlimited time. I think it is enough to save a to-do list.) and we add an option to link the database?

**Comment by OlegWock** [comment]
Yeah, that could be a possible solution. But this will be kind of ad-hoc solution, and I'm afraid might be quite confusing/complicated for non-tech users

I have a few features in mind that will benefit from having full-fledged back-end (like gmail or spotify integration). And thus I'm leaning into that side. But this is kinda depends on how successful my attempts at marketing will be, if extension stays at ~same audience as now I'll probably go with your proposed solution of using 3rd party database provider and asking user to register there. But if everything goes smooth, we'll go with companion back-end to provide sync and other features

**Comment by OlegWock** [comment]
I removed 'Pull requests are welcome' label as this feature is currently in a limbo and it will be more clear how to proceed after in a month or two

**Comment by sansmoraxz** [comment]
FYI [storage.sync](https://developer.chrome.com/docs/extensions/reference/storage/#property-sync) can help. Provided not storing a large payload.

**Comment by OlegWock** [comment]
I considered this, but it our payloads are larger that limits and we do update storage quite often

**Comment by karma1428** [comment]
.

**Comment by sansmoraxz** [comment]
I wonder will it really grow that much?

I have 6 different pages. My total backup size of exported configs is hardly even 12 kb.

Granted I am just stating out and was just looking for a `nightab` alternative while I stumbled upon this project.

**Comment by sansmoraxz** [comment]
Maybe some button to sync manually.

Checking for the size of the compressed payload and generating error messages if exceeded can be a good balance imo.

**Comment by OlegWock** [comment]
That could work for some (lean) setups, but it won't work for other users. My setup, for example, is 162kb compressed. Even if we workaround limitations of `chrome.sync` (8 kb per item, 100kb max), it won't be easy to code and maintain. I'm on 'good solution or nothing' side in this debate.

We might not need to code all sync logic from scratch and incorporate projects like [electric-sql](https://electric-sql.com/), [TinyBase](https://tinybase.org/) or [cr-sqlite](https://github.com/vlcn-io/cr-sqlite), but even in this case we'll need some kind of central server with user management, which is big chunk of work and maintenance.

**Comment by sansmoraxz** [comment]
Hey @OlegWock I gave this some thought. Truth be told I am not really a fan of having my extensions configs synced to somewhere I can't directly control.

What do you think of using Github to store the configs? Namely, just sync the user-generated PAT tokens and repo URLs. And then [directly update the files through Github's REST API](https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents).

**Comment by OlegWock** [comment]
@sansmoraxz If I would to implement sync, I'd prefer a solution which will work for majority of users (most of them aren't very tech-savy), so solution with github isn't ideal. 

However, if you (or any other person) are willing to implement sync with GitHub (or any other service which can be controlled by user, like their dropbox or something) and submit PR, I'll merge it (likely hidden under 'Advanced settings' or something like that).

From my side I'll be happy to answer any questions about code and guide you if you decide to give it a try.

**Comment by sansmoraxz** [comment]
I tried working on this, but the workflow is too damn slow. Not sure if it's specific to extension development, or just the huge webpack config. Dev server doesn't even work.

**Comment by sansmoraxz** [comment]
Anyway for anyone trying to take a dig at this with [octokit](https://github.com/octokit/octokit.js).

I made a small api server implementation.


For viewing contents of  file:
```js
import { Octokit } from "octokit";

export async function POST(request: Request) {
  // path is the expected properties in the request body
  const owner = process.env.GITHUB_OWNER || "octocat";
  const repo = process.env.GITHUB_REPO || "hello-world";

  console.log("request: ", request);
  const { path } = await request.json();

  const octokit = new Octokit({
    auth: process.env.GITHUB_TOKEN,
  });

  try {
    const x = await octokit.rest.repos.getContent({
      owner: owner,
      repo: repo,
      path: path,
      headers: {
        "X-GitHub-Api-Version": "2022-11-28",
      },
    });

    if (x.status !== 200 && x.status !== 201) {
      throw new Error("Failed to update file");
    }

    return new Response(
      JSON.stringify({
        status: 200,
        sha: x.data?.content?.sha,
      }),
      { status: 200 }
    );
  } catch (e) {
    console.error(e);
    return new Response(
      JSON.stringify({
        status: 500,
        sha: null,
      }),
      { status: 500 }
    );
  }
}
```

For updating the file you need to validate against the exisitng sha of the file. This should be available on the calling the read api. Sample here:

```js
import { Octokit } from "octokit";

export async function POST(request: Request) {
  // data, sha, and path are the expected properties in the request body
  const owner = process.env.GITHUB_OWNER || "octocat";
  const repo = process.env.GITHUB_REPO || "hello-world";

  console.log("request: ", request);
  const { data, sha, path } = await request.json();

  const octokit = new Octokit({
    auth: process.env.GITHUB_TOKEN,
  });

  try {
    const x = await octokit.rest.repos.createOrUpdateFileContents({
      owner: owner,
      repo: repo,
      path: path,
      sha: sha,
      message: "update",
      committer: {
        name: "github-actions",
        email: "41898282+github-actions[bot]@users.noreply.github.com",
      },
      content: Buffer.from(data).toString("base64"),
      headers: {
        "X-GitHub-Api-Version": "2022-11-28",
      },
    });

    if (x.status !== 200 && x.status !== 201) {
      throw new Error("Failed to update file");
    }

    return new Response(
      JSON.stringify({
        status: 200,
        sha: x.data.content?.sha,
      }),
      { status: 200 }
    );
  } catch (e) {
    console.error(e);
    return new Response(
      JSON.stringify({
        status: 500,
        sha: null,
      }),
      { status: 500 }
    );
  }
}
```

**Comment by OlegWock** [comment]
Didn't noticed that this was closed. I'm going to repopen it and keep it as 'main' issue for all cloud sync requests, as this one has the most actually useful info. If you don't want to get notifications, please unsubscribe from this issue by clicking button to the right.

**Comment by arknano** [comment]
Nothing very meaningful to add, but I just wanted to lodge that a lack of sync is definitely the biggest QOL issue I have as a daily user of Anori. I use the extension at work, which involves switching computers fairly regularly - I keep the archives on my work OneDrive and sync using that. It's a little tedious but not the end of the world. With that said, I don't really see how it can be solved in a way thats any easier than exporting/importing the archives (as all options would presumably require some kind of import/authentication step).

The only thought I have, and honestly I don't expect it to be viable/possible - is there a way to store the data using the browser(s) synchronisation features? e.g. logging into chrome with your google account, mozilla sync. 

Anyway, thanks for the work on the extension.

**Comment by OlegWock** [comment]
This has been released some time ago, so closing the issue. Find more info directly in the extension or on this page https://anori.app/plus
